### ISSUE TITLE
`fit_generator` and `train_on_batch` are not well documented and do not work with variable-size inputs

### ISSUE THREAD

**Issue by fvisin** [fchollet/keras#2539]
The format of the input that `fit_generator` expects to receive from the generator object is not clearly documented.

According to the documentation:

> The output of the generator must be either
>                 - a tuple (inputs, targets)
>                 - a tuple (inputs, targets, sample_weights).
>                 All arrays should contain the same number of samples. The generator is expected to loop over its data indefinitely. An epoch finishes when `samples_per_epoch` samples have been seen by the model.

The format of `inputs` and `targets` is not properly documented, but some assumptions are made in the code.
1. For some reason in case `x` is a list, Keras expects the _batches_ to be in the _second_ axis (see [L1359](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L1359)). It is not clear to me and is not documented what should go on the first axis, but I have reason to believe it has to do with `self.input_names`, whatever that is (see below). This comment https://github.com/fchollet/keras/issues/1428#issuecomment-180363196 is also related.
2. Later on, more undocumented assumptions on the data format are made. Specifically, this is the stacktrace that I reconstructed by manually (see #2538) inspecting the code:
   `fit_generator` --> `train_on_batch(x, ..)` --> `_standardize_user_data(x, ..)` --> `standardize_user_data(x, ..)`.
   
   This last function expects as an input a **list of arrays**, **dictionary of arrays**, or as a **single array**. 
   1. This is not documented in `fit_generator` and `train_on_batch`. 
   2. This does not work when the data is a list of lists (specifically, it will fail every time the `shape` argument is accessed (see e.g. [L91](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L91))
   3. As a consequence of ii., it is not possible to use `fit_generator` and `train_on_batch` with variable-size inputs (see example below).

**Example:**
Say that I have some 5D video data of dimensions [batches, frames, height, width, channels], where each batch will contain a 20 frames with the same resolution, but the resolution of the video varies among the batches. This cannot be saved as a numpy array because the data is not homogeneous. This breaks the assumption of `standardize_user_data` and makes it impossible to use Keras for this kind of data AFAIK.

**Comment by braingineer** [comment]
> The format of the input that fit_generator expects to receive from the generator object is not clearly documented.

I do agree with this a bit. I personally found that gandering at the code and testing revealed most things.  I've been using Keras for about 2 weeks and the 1.0 API is still hot of the presses (so to speak).  This is the price of bleeding edge.  

I'm not sure about a bunch of your issues but I can point you in the right direction for a couple. 

> For some reason in case x is a list, Keras expects the batches to be in the second axis (see L1359). It is not clear to me and is not documented what should go on the first axis

If you look at the rest of that code:
```python
                if type(x) is list:
                    batch_size = len(x[0])
                elif type(x) is dict:
                    batch_size = len(list(x.values())[0])
                else:
                    batch_size = len(x)
```

You'll see that if x isn't a list or a dict, it assumes x is the data.  If x is a list, it assumes that it is a list of data (aka, if you have [multiple inputs](http://keras.io/getting-started/functional-api-guide/#multi-input-and-multi-output-models)) .  I don't understand your point about what should go on the first axis.  Why would a list be considered an axis?  


> `_standardize_user_data`  -> `standardize_user_data`

where are you finding the second function?  I am getting as far as [here](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L898) where the only functions called are `standardize_input_data`, `standardize_sample_weights`,  `standardize_class_weights`, and `standardize_weights`.  


> This does not work when the data is a list of lists (specifically, it will fail every time the shape argument is accessed (see e.g. L91)

How did you make it past [this line](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L80) without a shape property? 


>Example

I don't work with video, so I don't know how common that use case is, but it seems you're the first to stumble across it.  Do you have a proposal for how to handle such data?  How would you like to see it go through?  

> shape [batches, frames, height, width, channels]

To make sure I understand it: 
X[0] is batch 0 and it has shape (f,h,w,c) and that may be different than the next 4-tuple in batch 2, but f always remains the same?   I am confused, then. why isn't len(X[0]) an adequate way to measure your batch size?   len(X[0])  in your example would always be 20.

**Comment by fvisin** [comment]
>I'm not sure about a bunch of your issues but I can point you in the right direction for a couple.

Thank you!

> You'll see that if x isn't a list or a dict, it assumes x is the data. If x is a list, it assumes that it is a list of data (aka, if you have multiple inputs) . 

I agree, in fact I only referred to the case where x is a list ("in case x is a list") :)
The "multiple inputs" explanation clarifies the data format though.

> I don't understand your point about what should go on the first axis. Why would a list be considered an axis?

It feels easier (and more natural, being used to numpy arrays) to refer to each list as an axis, but what I really meant is "the outer list".

> _standardize_user_data -> standardize_user_data
where are you finding the second function? I am getting as far as here where the only functions called are standardize_input_data, standardize_sample_weights, standardize_class_weights, and standardize_weights.

>Edit: I think you had meant standardize_input_data.

Yes, sorry! I got lazy and I copied and pasted without noticing it was not the same name. Sorry about that!

>>This does not work when the data is a list of lists (specifically, it will fail every time the shape argument is accessed (see e.g. L91)

>How did you make it past this line without a shape property?

In fact I didn't. :)
It fails every time `shape` is accessed, so yes it will fail at the line you refer to as well.

>>shape [batches, frames, height, width, channels]

>To make sure I understand it: 
X[0] is batch 0 and it has shape (f,h,w,c) and that may be different than the next 4-tuple in batch 2, but f always remains the same? I am confused, then. why isn't len(X[0]) an adequate way to measure your batch size? len(X[0]) in your example would always be 20.

>edit: I also misunderstood another thing (it's 4 am where I'm at.. sorry). Your first dimension is the batch index, so len(X[0]) would be 20 and not the number of samples in your batch, therefore False. There is an assumption that lists are only used to handle multiple inputs. Clearly it's an assumption that needs to be fixed. It seems like there would be some mucky bookkeeping to track all edge cases, though.

It's getting complicated, but I think we are on the same page! :) Thank you again for your help and sorry if I was not perfectly clear in some of the passages of my original message.

I confirm what you understood about the shape of the data in my example. 
`len(X[0])` works in my case if I modify my generator to return X incapsulated into a list (i.e. `yield [X]` instead of `yield X`) to "trick" Keras and go past the assumption that lists are only used to handle multiple inputs. In this case `len(X[0])` is the batch size - as Keras expects - and `X[0][0]` is batch 0 and it has shape (f,h,w,c). This will still fail [here](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L80), as you pointed out, because X[0] \(i.e. `array` in the code) is a list and not a numpy array.

> What if you assigned your input a name and used the dictionary style method of passing data? It might assume more delicate care with your data (and would at least pass the correct batch inference code correctly)

I will try it right away and report back, thank you for the idea.

> I don't work with video, so I don't know how common that use case is, but it seems you're the first to stumble across it. Do you have a proposal for how to handle such data? How would you like to see it go through?

I move this question at the end of the conversation to build on top of what I wrote so far. I think the best way to handle this would be to a) drop the assumption that lists are only used in case of multiple inputs b) handle lists by checking the type of the input everywhere you need to.

I don't know enough about Keras, but a) could probably be solved by using `names` to determine how many inputs to expect. [This](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L1358:L1359)  could become something like:
```python
if type(x) is list:
   batch_size = len(x[0]) if len(self.input_names) > 1 else len(x)
```
or, you could have something like this somewhere at the very beginning:
```python
if istype(x, (list, tuple)) and if len(self.input_names) > 1:
   x = [x]
```
To solve b), `standardize_input_data` (and probably many other functions) should be modified to handle numpy.array *and* lists.

**Comment by fvisin** [comment]
Quick update: the dictionary solution still fails [here](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L358). I begin to think there is no easy workaround. I will try to modify my experiments to work around Keras's assumptions, but I think this is a big limitation for Keras that should be addressed.

**Comment by fvisin** [comment]
Errata corrige: sorry, I meant [here](https://github.com/fchollet/keras/blob/3779b8a008d199fe1f2ea0b05254a7d775500e5a/keras/engine/training.py#L386). I updated the original comment.

**Comment by kureta** [comment]
I have a problem with `fit_generator`. If this does not belong here I can open a new issue. Here it goes:

I have set up my generator to print the number of current iteration and notify me after 66 iterations. My `batch_size` is 128. So my `samples_per_epoch`should be 66*128. But when I run this:

    self.model.fit_generator(generator=my_gen, samples_per_epoch=128*66, nb_epoch=1)

The output is this:

```
0
Epoch 1/1
 1
 2
 3
 4
 5
 6
 7
 8
 9
 128/8448 [..............................] - ETA: 7s - loss: -0.5599
 10
 11
 256/8448 [..............................] - ETA: 6s - loss: -0.5714
 12
 384/8448 [>.............................] - ETA: 6s - loss: -0.5453
 13
 512/8448 [>.............................] - ETA: 5s - loss: -0.5252
 14
 640/8448 [=>............................] - ETA: 5s - loss: -0.5159
 15
 768/8448 [=>............................] - ETA: 5s - loss: -0.5279
 16
 896/8448 [==>...........................] - ETA: 5s - loss: -0.5403
 17
1024/8448 [==>...........................] - ETA: 5s - loss: -0.5434
 18
1152/8448 [===>..........................] - ETA: 5s - loss: -0.5380
 19
1280/8448 [===>..........................] - ETA: 5s - loss: -0.5330
 20
1408/8448 [====>.........................] - ETA: 5s - loss: -0.5360
 21
1536/8448 [====>.........................] - ETA: 4s - loss: -0.5357
 22
1664/8448 [====>.........................] - ETA: 4s - loss: -0.5349
 23
1792/8448 [=====>........................] - ETA: 4s - loss: -0.5370
 24
1920/8448 [=====>........................] - ETA: 4s - loss: -0.5366
 25
2048/8448 [======>.......................] - ETA: 4s - loss: -0.5335
 26
2176/8448 [======>.......................] - ETA: 4s - loss: -0.5363
 27
2304/8448 [=======>......................] - ETA: 4s - loss: -0.5346
 28
2432/8448 [=======>......................] - ETA: 4s - loss: -0.5344
 29
2560/8448 [========>.....................] - ETA: 4s - loss: -0.5337
 30
2688/8448 [========>.....................] - ETA: 4s - loss: -0.5318
 31
2816/8448 [=========>....................] - ETA: 3s - loss: -0.5346
 32
2944/8448 [=========>....................] - ETA: 3s - loss: -0.5357
 33
3072/8448 [=========>....................] - ETA: 3s - loss: -0.5351
 34
3200/8448 [==========>...................] - ETA: 3s - loss: -0.5397
 35
3328/8448 [==========>...................] - ETA: 3s - loss: -0.5426
 36
3456/8448 [===========>..................] - ETA: 3s - loss: -0.5430
 37
3584/8448 [===========>..................] - ETA: 3s - loss: -0.5411
 38
3712/8448 [============>.................] - ETA: 3s - loss: -0.5409
 39
3840/8448 [============>.................] - ETA: 3s - loss: -0.5430
 40
3968/8448 [=============>................] - ETA: 3s - loss: -0.5430
 41
4096/8448 [=============>................] - ETA: 3s - loss: -0.5429
 42
4224/8448 [==============>...............] - ETA: 2s - loss: -0.5434
 43
4352/8448 [==============>...............] - ETA: 2s - loss: -0.5443
 44
4480/8448 [==============>...............] - ETA: 2s - loss: -0.5436
 45
4608/8448 [===============>..............] - ETA: 2s - loss: -0.5439
 46
4736/8448 [===============>..............] - ETA: 2s - loss: -0.5458
 47
4864/8448 [================>.............] - ETA: 2s - loss: -0.5472
 48
4992/8448 [================>.............] - ETA: 2s - loss: -0.5464
 49
5120/8448 [=================>............] - ETA: 2s - loss: -0.5467
 50
5248/8448 [=================>............] - ETA: 2s - loss: -0.5462
 51
5376/8448 [==================>...........] - ETA: 2s - loss: -0.5468
 52
5504/8448 [==================>...........] - ETA: 2s - loss: -0.5473
 53
5632/8448 [===================>..........] - ETA: 1s - loss: -0.5483
 54
5760/8448 [===================>..........] - ETA: 1s - loss: -0.5480
 55
5888/8448 [===================>..........] - ETA: 1s - loss: -0.5497
 56
6016/8448 [====================>.........] - ETA: 1s - loss: -0.5495
 57
6144/8448 [====================>.........] - ETA: 1s - loss: -0.5507
 58
6272/8448 [=====================>........] - ETA: 1s - loss: -0.5518
 59
6400/8448 [=====================>........] - ETA: 1s - loss: -0.5533
 60
6528/8448 [======================>.......] - ETA: 1s - loss: -0.5548
 61
6656/8448 [======================>.......] - ETA: 1s - loss: -0.5555
 62
6784/8448 [=======================>......] - ETA: 1s - loss: -0.5561
 63
6912/8448 [=======================>......] - ETA: 1s - loss: -0.5583
 64
7040/8448 [========================>.....] - ETA: 0s - loss: -0.5593
 65
7168/8448 [========================>.....] - ETA: 0s - loss: -0.5601
FINISH!!!

 66
7296/8448 [========================>.....] - ETA: 0s - loss: -0.5605
 67
7424/8448 [=========================>....] - ETA: 0s - loss: -0.5612
 68
7552/8448 [=========================>....] - ETA: 0s - loss: -0.5613
 69
7680/8448 [==========================>...] - ETA: 0s - loss: -0.5617
 70
7808/8448 [==========================>...] - ETA: 0s - loss: -0.5607
 71
7936/8448 [===========================>..] - ETA: 0s - loss: -0.5603
 72
8064/8448 [===========================>..] - ETA: 0s - loss: -0.5596
 73
8192/8448 [============================>.] - ETA: 0s - loss: -0.5597
 74
8320/8448 [============================>.] - ETA: 0s - loss: -0.5610
 75
8448/8448 [==============================] - 5s - loss: -0.5603
```

It seems that first 10 yields were skipped. What might the problem be?

**Comment by swirlingsand** [comment]
One example I think that would help people with the fit_generator documentation is something like:

samples_per_epoch is the **total** number of samples per epoch.

For example, if samples_per_epoch is set to 50, and you generate 10 images per call from a generator function, fit_generator will call your generator 5 times.

More specifically, if your generator function yields a tensor such as [5, 32, 32, 3], (perhaps representing a a stack of 5, 32x32 images, with 3 colour channels.), and your samples_per_epoch is set to 50, your generator function will be called 5 times.

**Comment by patyork** [comment]
I don't know how it could be more clear. "An epoch finishes when total_samples_per_epoch samples have been seen by the model." as opposed to "An epoch finishes when samples_per_epoch samples have been seen by the model." seems fully redundant.

**Comment by swirlingsand** [comment]
Agree. I think that sentence says the same thing - the current one is better worded too.

It's more the part about batches. Since it's possible to feed a [1, 32, 32, 3] tensor, it wasn't clear (at least to me), that the "samples" is in reference to the x value in [x, 32, 32, 3], and not the entire tensor itself.

I'm a beginner at this so I'm probably not the best person to write the wording - just wanted to call attention to an opportunity to maybe clarify and perhaps offer a wording myself to help get the ball started.
